packages/@aws-cdk/aws-s3-notifications/lib/lambda.ts

```diff
@@ -11,19 +11,28 @@ export class LambdaDestination implements s3.IBucketNotificationDestination {
   }
 
   public bind(_scope: Construct, bucket: s3.IBucket): s3.BucketNotificationDestinationConfig {
-    const permissionId = `AllowBucketNotificationsFrom${bucket.node.uniqueId}`;
+    const permissionId = `AllowBucketNotificationsTo${this.fn.permissionsNode.uniqueId}`;
 
-    if (this.fn.permissionsNode.tryFindChild(permissionId) === undefined) {
+    if (!Construct.isConstruct(bucket)) {
+      throw new Error(`LambdaDestination for function ${this.fn.permissionsNode.uniqueId} can only be configured on a
+        bucket construct (Bucket ${bucket.bucketName})`);
+    }
+
+    if (bucket.node.tryFindChild(permissionId) === undefined) {
       this.fn.addPermission(permissionId, {
         sourceAccount: Stack.of(bucket).account,
         principal: new iam.ServicePrincipal('s3.amazonaws.com'),
         sourceArn: bucket.bucketArn,
+        // the bucket stack already has a dependency on the lambda stack because
+        // of the notification resource. therefore adding the permission to the lambda
+        // stack would create a circular dependency, so we add it to the bucket scope.
```

Comment:
Minor: for someone reading this outside of the context of this PR will be confused what the lambda stack and bucket stacks are.

Something like this maybe?

```suggestion
        // Placing the permissions node in the same scope as the s3 bucket.
        // Otherwise, there is a circular dependency when the s3 bucket
        // and lambda functions declared in different stacks.
```

Reply:
👍